File: modules/libs/core/usecase/flashcards/session.go
package flashcards

import (
	"context"
	"errors"
	"time"

	"github.com/jiva-studio/numen/modules/libs/core/domain"
	"github.com/jiva-studio/numen/modules/libs/core/flashcards/review"
	"github.com/jiva-studio/numen/modules/libs/core/port"
)

// ErrBothNamed is a session named a deck and a preset at once. Which cards were
// meant is a question, and it is put back to the caller.
var ErrBothNamed = errors.New("a session is opened over one deck or over one preset")

// ErrSchedulesNothing is a preset with nothing to ask in the day being sat. The
// message says which of the reasons it is, in the person's own words.
var ErrSchedulesNothing = errors.New("this preset schedules nothing today")

// Scope is what a session is opened over: every deck the vault holds, one of
// its decks, or one of its presets.
//
// The whole vault is the ordinary way to sit down to this: a person owes what
// they owe, and which file a card is written in is not something they think
// about.
type Scope struct {
	// Deck is the path of one deck.
	Deck string
	// Preset is the note one preset stands in. The preset scheduling the decks
	// naming none stands in no note, so Named says a preset was named at all.
	Preset string
	Named  bool
}

// OverDeck is a session over one deck.
func OverDeck(path string) Scope { return Scope{Deck: path} }

// ByPreset is a session over the cards of every deck pointing at one preset,
// held to that preset's budget.
func ByPreset(preset string) Scope { return Scope{Preset: preset, Named: true} }

// QueuedCardFace is one card face as the session queues it: where it stands,
// how it is laid out, and where the answers so far have left it.
type QueuedCardFace struct {
	CardFace
	Schedule review.Schedule
	// Ahead is how long each of the four answers would leave this card, from
	// the moment it is asked. A person choosing between them is choosing
	// between these, so they are worked out with the card and not after it.
	Ahead map[review.Rating]time.Duration
}

// SessionResult is what a person sits down to: the cards to ask, and what could
// not be acted on in getting them.
type SessionResult struct {
	// Queue is what to ask, in the order to ask it.
	Queue []QueuedCardFace
	// Unwritten are the decks holding a card with no mark that could not be
	// given one. Their cards are not in the queue and are asked for at the next
	// session.
	Unwritten []string
	// Skipped is how many lines of the vault's answers could not be read: a run
	// that stopped partway, or a line of a version this build does not know.
	Skipped int
}

// Session is what a person is asked, in the order they are asked it.
//
// A card owed and answered before comes first, the one waiting longest at the
// front, because a card left late is the one closest to being forgotten. Cards
// nobody has answered come after them, in the order they stand in their decks:
// a person wrote them in an order, and it is as good an order as any.
//
// Each deck is held to the budget its preset keeps today, and what was already
// answered today is off that budget.
type Session struct {
	// Marks gives a mark to the cards of this vault that carry none, so that
	// what is asked can be answered. It is the one write flashcards makes, and it
	// is made when a person sits down to a vault.
	Marks     MarkCards
	CardFaces ListCardFaces
	Schedules Schedules
	// Presets says which preset each deck is scheduled by. A build holding no
	// links schedules every deck by the defaults.
	Presets Presets
	Day     review.Day
	Now     port.Clock
}

// NewSession is what a person sits down to their cards through: what gives the
// unmarked cards their marks, what stands in the vault, where the answers have
// left each card face, which preset each deck is scheduled by, where one day of
// review gives way to the next, and what time it is.
func NewSession(
	marks MarkCards, faces ListCardFaces, schedules Schedules, presets Presets,
	day review.Day, now port.Clock,
) Session {
	return Session{
		Marks: marks, CardFaces: faces, Schedules: schedules, Presets: presets,
		Day: day, Now: now,
	}
}

// Execute is what to ask, in order.
//
// Scope is the deck or the preset the session is opened over. Naming both is
// ErrBothNamed, and a preset with nothing to ask today is ErrSchedulesNothing
// with the reason.
func (u Session) Execute(
	ctx context.Context, v domain.Vault, over Scope,
) (SessionResult, error) {
	if over.Named && over.Deck != "" {
		return SessionResult{}, ErrBothNamed
	}
	marked, err := u.Marks.Execute(ctx, v)
	if err != nil {
		return SessionResult{}, err
	}
	faces, err := u.CardFaces.Execute(ctx, v)
	if err != nil {
		return SessionResult{}, err
	}

	// The log is read once here and the schedules worked out from it, so that
	// what a person is told about lines that could not be read is the reading
	// their own cards were laid out from.
	held, err := Log{Stores: u.Schedules.Logs}.Read(ctx, v)
	if err != nil {
		return SessionResult{}, err
	}
	// One reading of this vault's presets answers both the schedulers the cards
	// are worked out by and the budgets they are held to.
	reading := u.Presets.Reading()
	asks, err := u.Schedules.under(ctx, v, reading, faces)
	if err != nil {
		return SessionResult{}, err
	}
	schedules := u.Schedules.getSchedulesCached(ctx, v, held, asks)

	now := u.Now()
	day, err := getBudgets(
		ctx, v, reading, u.Day, faces, schedules, held,
		u.Schedules.By, u.Schedules.at, now,
	)
	if err != nil {
		return SessionResult{}, err
	}
	holds := day.asks(faces, schedules, u.Day, now, over)
	if over.Named && len(holds.seen)+len(holds.fresh) == 0 {
		return SessionResult{}, day.refuses(over.Preset)
	}

	// How loaded each day of review already is, which is what a card put on one
	// of them is weighed against.
	on := review.NewDueByDay(u.Day)
	for _, s := range schedules {
		on.Holds(s.Due)
	}

	out := SessionResult{Unwritten: marked.Unwritten, Skipped: held.Skipped}
	out.Queue = make([]QueuedCardFace, 0, len(holds.seen)+len(holds.fresh))
	for _, one := range holds.seen {
		s := schedules[one.ID]
		out.Queue = append(out.Queue, QueuedCardFace{
			CardFace: one, Schedule: s, Ahead: ahead(asks.under, on, one.ID, s, now),
		})
	}
	for _, one := range holds.fresh {
		out.Queue = append(out.Queue, QueuedCardFace{
			CardFace: one,
			Ahead:    ahead(asks.under, on, one.ID, review.Schedule{}, now),
		})
	}
	return out, nil
}

// ahead is how long each of the four would leave a card standing where this
// schedule leaves it. The card is left where it was, and so is the table of how
// loaded each day is: only the answer a person gives lands anywhere.
//
// The scheduler is the one this card face is scheduled by and the placement is
// its own preset's, so the window under each button is the day the card will
// come back on.
func ahead(
	under review.Assignment, on *review.DueByDay, face review.CardFaceID,
	s review.Schedule, now time.Time,
) map[review.Rating]time.Duration {
	if under == nil {
		return nil
	}
	one := under(face)
	if one.By == nil {
		return nil
	}
	out := make(map[review.Rating]time.Duration, 4)
	for _, r := range []review.Rating{review.Again, review.Hard, review.Good, review.Easy} {
		due := one.By.Next(s, now, r).Due
		out[r] = one.Preset.Lands(on, now, due).Sub(now)
	}
	return out
}

File: modules/apps/desktop/flashcards/src/session.ts
/**
 * One sitting: the cards left to ask, where in them a person is, and what has
 * been written into the vault.
 *
 * Apart from the template because these are the rules that decide what a person
 * is asked and what their vault is told, and a rule inside a component is a rule
 * only exercised by looking at the screen.
 */
import { computed, ref } from 'vue'
import type { Ref } from 'vue'

import { rated } from './core'
import type { CardFace, Grade, Intervals } from './core'

/** What a sitting asks of the application, and no more of it than that. */
export interface SessionClient {
  /**
   * Deck is one deck, or empty for every deck the vault holds. Preset holds it
   * to the decks one preset schedules, and to the budget that preset keeps.
   */
  startSession(said: { vault: string; deck: string; preset?: string }): Promise<SessionStart>
  answerCard(said: {
    vault: string
    run: string
    card: string
    face: string
    rating: number
    tookMs: bigint
  }): Promise<{ answer: string }>
  takeBackAnswer(said: { vault: string; run: string; answer: string }): Promise<unknown>
}

/** What opening a sitting comes back with. */
export interface SessionStart {
  run: string
  asked: readonly {
    deck: string
    section: string
    card: string
    face: string
    heading: string
    front: string
    back: string
    seen: boolean
    ahead?: { again: bigint; hard: bigint; good: bigint; easy: bigint } | undefined
  }[]
  unwritten: readonly string[]
  skipped: number
}

/** What a sitting could not act on, for the window to say once as it opens. */
export interface Report {
  /** The decks holding a card that could not be given a mark. */
  readonly unwritten: readonly string[]
  /** How many lines of the vault's answers could not be read. */
  readonly skipped: number
}

/** What a sitting is built over: the application, and what it says went wrong. */
export interface SessionDeps {
  cards: SessionClient
  failed(why: unknown): void
  /** When it is, in milliseconds. How long a card stood there is measured with it. */
  now?(): number
}

export function session(deps: SessionDeps) {
  const now = deps.now ?? (() => Date.now())

  const vault = ref('')
  const run = ref('')
  const asked = ref<readonly CardFace[]>([])
  const at = ref(0)
  const shown = ref(false)
  const answers = ref<string[]>([])

  /**
   * Whether an answer is on its way to the vault. A card is answered once: the
   * line is written before the next card is put up, and a second press while
   * that is happening would write the same card twice and skip the one after
   * it.
   */
  const writing = ref(false)

  /** When the card now in front of the person was put there. */
  let put = now()

  const card = computed<CardFace | null>(() => asked.value[at.value] ?? null)
  const left = computed(() => asked.value.length - at.value)
  const over = computed(() => card.value === null)

  /** How many answers this sitting has written, which is what stands at the end. */
  const done = computed(() => answers.value.length)

  /**
   * The sitting let go of. What was answered is in the vault, and the next
   * sitting is opened whole: a card kept here is one a screen could fall back
   * to showing with nothing able to answer it.
   */
  const forget = () => {
    asked.value = []
    at.value = 0
    shown.value = false
    answers.value = []
    run.value = ''
  }

  /**
   * Sit down to a vault, or to one deck of it. What comes back is what the
   * sitting could not act on, and nothing when it could not be opened at all.
   */
  /**
   * Preset is the note one preset stands in, and the empty path is the preset
   * that schedules the decks naming none. Naming none at all sits to the deck.
   */
  const start = async (
    named: string,
    deck: string,
    preset?: string,
  ): Promise<Report | null> => {
    try {
      const opened = await deps.cards.startSession(
        preset === undefined
          ? { vault: named, deck }
          : { vault: named, deck, preset },
      )
      vault.value = named
      run.value = opened.run
      asked.value = opened.asked.map(asking)
      at.value = 0
      shown.value = false
      answers.value = []
      put = now()
      return { unwritten: opened.unwritten, skipped: opened.skipped }
    } catch (why) {
      deps.failed(why)
      return null
    }
  }

  const show = () => {
    shown.value = true
  }

  /**
   * Answer the card in front of the person. An answer that could not be written
   * leaves the card where it was: what a person said is theirs, and a card
   * moved past with nothing recorded is an answer lost.
   */
  const answer = async (how: Grade) => {
    const one = card.value
    if (!one || !shown.value || writing.value) return
    const took = now() - put
    writing.value = true
    try {
      const given = await deps.cards.answerCard({
        vault: vault.value,
        run: run.value,
        card: one.card,
        face: one.face,
        rating: rated[how],
        tookMs: BigInt(took),
      })
      answers.value.push(given.answer)
    } catch (why) {
      deps.failed(why)
      return
    } finally {
      writing.value = false
    }
    at.value += 1
    shown.value = false
    put = now()
  }

  /**
   * Take the last answer back. The card comes back with its answer showing,
   * which is where the person was when they pressed the wrong key.
   */
  const takeBack = async () => {
    const last = answers.value[answers.value.length - 1]
    if (last === undefined || writing.value) return
    writing.value = true
    try {
      await deps.cards.takeBackAnswer({ vault: vault.value, run: run.value, answer: last })
    } catch (why) {
      deps.failed(why)
      return
    } finally {
      writing.value = false
    }
    answers.value.pop()
    at.value = Math.max(0, at.value - 1)
    shown.value = true
    put = now()
  }

  return {
    vault: vault as Ref<string>,
    run,
    asked,
    at,
    shown,
    answers,
    writing,
    card,
    left,
    over,
    done,
    start,
    show,
    answer,
    takeBack,
    forget,
  }
}

/** One card as the window holds it: the seconds come across as numbers. */
const asking = (one: SessionStart['asked'][number]): CardFace => ({
  deck: one.deck,
  section: one.section,
  card: one.card,
  face: one.face,
  heading: one.heading,
  front: one.front,
  back: one.back,
  seen: one.seen,
  ahead: one.ahead
    ? ({
        again: Number(one.ahead.again),
        hard: Number(one.ahead.hard),
        good: Number(one.ahead.good),
        easy: Number(one.ahead.easy),
      } satisfies Intervals)
    : null,
})
